### PR TITLE
Added Interface for a builder of Actions

### PR DESCRIPTION
**Pull request by turcato-niccolo**
Written on the hypothesis that a class implementing this interface builds the action calculating a random 32-bit integer ID for the action.

### CODE REVIEW

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     * Hypothesis: this type of responses (ANSW) is built starting from the request (mainly to retrieve the action ID/number)
+     *
+     * QUESTION: is it good?
+     */
+    T buildPingAnsw(T request);
+
+
+    T buildInvite(P peer);
+
+    T buildInviteAnsw(T request, boolean accept);
```
Comment (ricdezen) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java]: boolean accepted would sound better

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+
+
+    T[] buildStore(P peer, P peerToStore);
```
Comment (ricdezen) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java]: Why does this return an array? Case 1 of store works just like find node and value. If you find a good way to overload every method with the array counterpart I'd be fine with it (Maybe implicit arrays like P... would work, but I know for a fact they can be troublesome with generics, the issue should disappear once we implement the class with the actual types)

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    T[] buildStoreAnsw(T request, P peer);
+
+    T buildStoreAnsw(T request, boolean confirmStore);
+
+
+
+    T buildFindNode(P peer, N node);
+
+    T[] buildFindNodeAnsw(T request, N[] nodes);
```
Comment (ricdezen) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java]: Why does find Node answer take in Nodes but other answer Peers? We should decide on one, the fact our PeerNodes extend both Peer and Node is a coincidence we shouldn't consider in the interface

**Comment by ricdezen** [comment]
Good thing you used Requests as parameters for Responses, I think I told someone sometime ago but forgot to add it in the paper

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java (hunk 3 above)
Comment (turcato-niccolo) [reply distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java]: Yeah, buildFindNodeAnsw should respond with P[] peers

**Comment by turcato-niccolo** [comment]
> Good thing you used Requests as parameters for Responses, I think I told someone some time ago but forgot to add it in the paper

Did some changes, is this more like it?

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public interface ActionsBuilder<T, P extends Peer, R extends Resource, N extends Node> {
+
+    T buildPing(P peer);
+
+
+    /**
+     * Hypothesis: this type of responses (ANSW) is built starting from the request (mainly to retrieve the action ID/number)
+     *
+     * QUESTION: is it good?
```
Comment (ricdezen) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java]: In general, yes, it's very unlikely you'll need to build a Response without the Request. Outside of testing at least. You read the Peer for an incoming Action as the Peer for the SMS, which is exactly who you wanna answer to.

**Review by ricdezen — APPROVED** [review]
Looks legit, don't forget specs before code review tho
P.S. Code review as in the course's Code Review, not necessarily ours

**Comment by turcato-niccolo** [comment]
> Looks legit, don't forget specs before code review tho
> P.S. Code review as in the course's Code Review, not necessarily ours

Nice, thanks, I'll add them right away

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java (hunk 4 above)
Comment (turcato-niccolo) [reply distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java]: As I thought👍

**Comment by turcato-niccolo** [comment]
In the previous commits, I forgot `T extends NodeActionStructure`

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+     */
+    T buildStoreAnsw(T request, boolean confirmStore);
+
+
+    /**
+     * @param peer the peer that will receive the built action command
+     * @param node the node of which find its address (on the used network)
+     * @return an action built to execute the find node command to the peer
+     */
+    T buildFindNode(P peer, N node);
```
Comment (kummy98) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java:74]: I don't understand why you use P peer as parameter in this method

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * @param request the received find node request from a peer
+     * @param peers the peers that are closest to the searched node (or the searched node's corresponding peer)
+     * @return an array of responses (one for each peer in peers) for the received request
+     */
+    T[] buildFindNodeAnsw(T request, P[] peers);
+
+
+    /**
+     * @param peer he peer that will receive the built action command
```
Comment (kummy98) [inline distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java]: I think here is the , and same here I don't understand why is it necessary to have the parameter peer

File: distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java (hunk 5 above)
Comment (turcato-niccolo) [reply distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/ActionsBuilder.java:74]: In the comment above the class, I've specified that the type P is a "physical" type of peer to which you forward the actions.
Classes KadAction and Network actions are valid when a Peer (SmsPeer in these cases) has been set. So to build a proper action you need a Peer

**Review by bortoletti-giorgia — APPROVED** [review]

**Pull request merged**